### PR TITLE
[Merged by Bors] - feat(FieldTheory/SeparableDegree): some results on separable degree one polynomials

### PR DESCRIPTION
**Pull request by acmepjz**
- `perfectField_iff_splits_of_natSepDegree_eq_one`: A field is perfect if and only if every separable degree one polynomial splits.
- add some criteria of a monic polynomial being separable degree one
- add some convenient results of separable degree of a polynomial

---

- [x] depends on: #10571
- [x] depends on: #10572

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

File: Mathlib/FieldTheory/SeparableDegree.lean [hunk 1]
```diff
@@ -705,3 +705,21 @@ theorem separable_inv (x : E) (hx : (minpoly F x).Separable) : (minpoly F x⁻¹
   separable_of_mem_isSeparable F E <| F⟮x⟯.inv_mem <| mem_adjoin_simple_self F x
 
 end Field
+
+/-- A field is a perfect field (which means that any irreducible polynomial is separable)
+if and only if every separable degree one polynomial splits. -/
+theorem perfectField_iff_splits_of_natSepDegree_eq_one (F : Type*) [Field F] :
+    PerfectField F ↔ ∀ f : F[X], f.natSepDegree = 1 → f.Splits (RingHom.id F) := by
```
Comment (alreadydone) [inline Mathlib/FieldTheory/SeparableDegree.lean:712]: We have lemmas about `f.natSepDegree = 1` assuming `Irreducible f` ([Irreducible.natSepDegree_eq_one_iff_of_monic'](https://leanprover-community.github.io/mathlib4_docs/Mathlib/FieldTheory/SeparableDegree.html#Irreducible.natSepDegree_eq_one_iff_of_monic') etc.), but without that assumption we should still be able to show `f` has the form `(X ^ p ^ n - C a) ^ m` (and you can require that `a : F` has no `p`th root in F).

Comment (acmepjz) [reply Mathlib/FieldTheory/SeparableDegree.lean:712]: I should say in the proof of this result I cheated a bit and uses `PerfectRing.ofSurjective`. The criterion of a monic polynomial being separable degree one is not used.

Comment (acmepjz) [reply Mathlib/FieldTheory/SeparableDegree.lean:886]: I have added the result you mentioned. The proof is very ugly though; improvements are welcome. Added a lot of intermediate results to various files. There is still an intermediate result `test1` which has no good descriptive names.

**Review by alreadydone — COMMENTED** [review]
Thanks for the effort! The results you prove are quite general and strong. As for the original result I suggested, it might be more direct to reuse [Irreducible.natSepDegree_eq_one_iff_of_monic](https://leanprover-community.github.io/mathlib4_docs/Mathlib/FieldTheory/SeparableDegree.html#Irreducible.natSepDegree_eq_one_iff_of_monic). Just extract a maximum power of an irreducible factor of the polynomial using [UniqueFactorizationMonoid.max_power_factor](https://leanprover-community.github.io/mathlib4_docs/Mathlib/RingTheory/UniqueFactorizationDomain.html#UniqueFactorizationMonoid.max_power_factor) (note: many unncessary hypotheses), and what's left is coprime to the irreducible factor by [Irreducible.coprime_or_dvd](https://leanprover-community.github.io/mathlib4_docs/Mathlib/RingTheory/PrincipalIdealDomain.html#Irreducible.coprime_or_dvd), so it can't share any common root with the irreducible factor (in any extension), and therefore must be constant (1 if monic) if the original polynomial has only one root. So the original polynomial is a power of the irreducible factor, which has the form X^q^n-a.

**Comment by acmepjz** [comment]
> it might be more direct to reuse [Irreducible.natSepDegree_eq_one_iff_of_monic](https://leanprover-community.github.io/mathlib4_docs/Mathlib/FieldTheory/SeparableDegree.html#Irreducible.natSepDegree_eq_one_iff_of_monic).

This was my unsuccessful attempt, seems that this proof is no shorter than current proof, though.

> Just extract a maximum power of an irreducible factor of the polynomial using [UniqueFactorizationMonoid.max_power_factor](https://leanprover-community.github.io/mathlib4_docs/Mathlib/RingTheory/UniqueFactorizationDomain.html#UniqueFactorizationMonoid.max_power_factor) (note: many unncessary hypotheses), and what's left is coprime to the irreducible factor by [Irreducible.coprime_or_dvd](https://leanprover-community.github.io/mathlib4_docs/Mathlib/RingTheory/PrincipalIdealDomain.html#Irreducible.coprime_or_dvd), so it can't share any common root with the irreducible factor (in any extension), and therefore must be constant (1 if monic) if the original polynomial has only one root. So the original polynomial is a power of the irreducible factor, which has the form X^q^n-a.

I was not aware of this approach! Seems that using [multiplicity](https://leanprover-community.github.io/mathlib4_docs/Mathlib/RingTheory/Multiplicity.html#multiplicity) is better?

By the way, `UniqueFactorizationMonoid.max_power_factor` requires the (base) ring to be `NormalizationMonoid`, which I think that a field should be a `NormalizationMonoid`. But currently mathlib has no instance for this yet.

**Comment by alreadydone** [comment]
> I was not aware of this approach! Seems that using [multiplicity](https://leanprover-community.github.io/mathlib4_docs/Mathlib/RingTheory/Multiplicity.html#multiplicity) is better?

That can work but is probably less convenient. I'm not sure in what generality you can show a polynomial have finite multiplicity in another, but at least a monic polynomial of positive degree has finite multiplicity in any nonzero polynomial.

> By the way, UniqueFactorizationMonoid.max_power_factor requires the (base) ring to be NormalizationMonoid, which I think that a field should be a NormalizationMonoid. But currently mathlib has no instance for this yet.

I think you need the polynomial ring to be a NormalizationMonoid, not the base ring, which you can get using Classical.arbitrary, because there's [instNonemptyNormalizedGCDMonoid_1](https://leanprover-community.github.io/mathlib4_docs/Mathlib/RingTheory/UniqueFactorizationDomain.html#instNonemptyNormalizedGCDMonoid_1) I just added. It may be hard to remove the condition from the lemma, since the lemma may be needed to prove the instance ...

**Comment by acmepjz** [comment]
> I'm not sure in what generality you can show a polynomial have finite multiplicity in another

All results in this files are polynomials over fields. In this case the polynomial ring itself is a PID, so I think the finiteness of the multiplicity should comes from the general results of multiplicity being finite for PIDs (or perhaps UFDs, or even [WfDvdMonoid](https://leanprover-community.github.io/mathlib4_docs/Mathlib/RingTheory/UniqueFactorizationDomain.html#WfDvdMonoid)). We should already have these results in mathlib, no?

**Comment by alreadydone** [comment]
Yeah, but I think max_power_factor can also be generalized to arbitrary UFD. The [Nontrivial R] [NormalizationMonoid R] [dec_dvd : DecidableRel Dvd.dvd] assumptions can all be removed once you have the NonemptyNormalizedGCD instance.

**Comment by alreadydone** [comment]
well, subtraction requires Ring because `sub` isn't a field of Semiring ...

**Comment by acmepjz** [comment]
> Yeah, but I think max_power_factor can also be generalized to arbitrary UFD. The [Nontrivial R] [NormalizationMonoid R] [dec_dvd : DecidableRel Dvd.dvd] assumptions can all be removed once you have the NonemptyNormalizedGCD instance.

In fact I don't know why `NormalizationMonoid` is required to prove this result. Also, I think the `n` in this result is just the `multiplicity`, in order to make the API consistent, I think this result should be rephrased using `multiplicity.Finite`.

**Comment by alreadydone** [comment]
The `multiplicity.finite` result is [multiplicity.exists_eq_pow_mul_and_not_dvd](https://leanprover-community.github.io/mathlib4_docs/Mathlib/RingTheory/Multiplicity.html#multiplicity.exists_eq_pow_mul_and_not_dvd) and max_power_factor could be derived from it, yes. Do we have that a non-unit in a WfDvdMonoid has finite multiplicity in any nonzero element?

**Comment by alreadydone** [comment]
It turns out the generalization of `max_power_factor` doesn't require cancellation, but the generalization of `multiplicity.finite_prime_left` does (otherwise you get examples where `a` is an idempotent different from 0 and 1). They do both generalize to non-units:
```
import Mathlib.RingTheory.UniqueFactorizationDomain

theorem max_power_factor {R} [CommMonoidWithZero R] [WfDvdMonoid R]
    {a b : R} (ha : ¬ IsUnit a) (hb : b ≠ 0) :
    ∃ (n : ℕ) (c : R), ¬a ∣ c ∧ b = a ^ n * c := by
  obtain ⟨c, ⟨n, rfl⟩, min⟩ := wellFounded_dvdNotUnit.has_min
    {c | ∃ n, a ^ n * c = b} ⟨b, 0, by rw [pow_zero, one_mul]⟩
  refine ⟨n, c, ?_, rfl⟩; rintro ⟨d, rfl⟩
  exact min d ⟨n + 1, by rw [pow_succ', mul_assoc]⟩
    ⟨(right_ne_zero_of_mul <| right_ne_zero_of_mul hb), a, ha, mul_comm _ _⟩

theorem multiplicity.finite_of_not_isUnit {R} [CancelCommMonoidWithZero R] [WfDvdMonoid R]
    {a b : R} (ha : ¬ IsUnit a) (hb : b ≠ 0) : multiplicity.Finite a b := by
  obtain ⟨n, c, ndvd, rfl⟩ := max_power_factor ha hb
  exact ⟨n, by rwa [pow_succ', mul_dvd_mul_iff_left (left_ne_zero_of_mul hb)]⟩
```

**Comment by acmepjz** [comment]
> It turns out the generalization of `max_power_factor` doesn't require cancellation, but the generalization of `multiplicity.finite_prime_left` does (otherwise you get examples where `a` is an idempotent different from 0 and 1).

Let me see. Do you mean that although `¬a ∣ c ∧ b = a ^ n * c`,  but it's still possible that `a ^ (n + 1) ∣ b` since it's not a domain?

>They do both generalize to non-units

Great! Waiting for your PRs.

**Comment by alreadydone** [comment]
Yes, that's what I mean. I have a deadline to meet and won't open a PR today, so feel free to.

**Comment by acmepjz** [comment]
> I have a deadline to meet and won't open a PR today, so feel free to.

No worries. Currently I don't need to use such generalization.

**Review by alreadydone — APPROVED** [review]
LGTM, thanks!

**Comment by riccardobrasca** [comment]
Thanks!

bors merge

**Pull request closed without merging**